lib/api-helper/goal/mock.ts

```diff
@@ -0,0 +1,127 @@
+import { SoftwareDeliveryMachineConfiguration } from "../../api/machine/SoftwareDeliveryMachineOptions";
+
+/**
+ * Enum to describe goal sizes when creating mock goal executors
+ */
+export enum MockGoalSize {
+    Small = 5,
+    Medium = 20,
+    Large = 60,
+    ExtraLarge = 90,
+}
+
+/**
+ * Options to configure mocks for one particular goal
+ */
+export interface MockGoal {
+    goal: Goal;
+    size: MockGoalSize | number;
+    result?: number;
+    mock?: ExecuteGoal;
+}
+
+/**
+ * Options for the mock goal executor support
+ */
+export interface MockOptions {
+    enabled: boolean | ((goal: SdmGoalEvent) => boolean);
+    defaultSize?: MockGoalSize | number;
+    goals?: MockGoal[];
+    randomBy?: number;
+}
+
+/**
+ * Default mock options to be merged into custom options provided on configuration.sdm.mock
+ */
+export const DefaultMockOptions: MockOptions = {
+    enabled: false,
+    defaultSize: MockGoalSize.Medium,
+    goals: [],
+    randomBy: 0.2,
+};
+
+/**
+ * Create mock goal executor for a certain goal
+ * @param goal
+ * @param configuration
+ */
+export function mockGoalExecutor(goal: Goal,
+                                 sdmGoal: SdmGoalEvent,
+                                 configuration: SoftwareDeliveryMachineConfiguration): ExecuteGoal | undefined {
+    const options = _.merge(DefaultMockOptions, _.get(configuration, "sdm.mock", {})) as MockOptions;
+
+    let enabled;
+    if (typeof options.enabled === "boolean") {
+        enabled = options.enabled;
+    } else {
+        enabled = options.enabled(sdmGoal);
+    }
+
+    if (enabled) {
+        const mock = options.goals.find(g => g.goal === goal);
+        if (mock) {
+            if (mock.mock) {
+                return mock.mock;
+            } else {
+                return createGoalExecutor(randomize(mock.size, options.randomBy), mock.result);
+            }
+        } else {
+            return createGoalExecutor(randomize(options.defaultSize, options.randomBy));
+        }
+    }
+
+    return undefined;
+}
+
+export function randomize(size: MockGoalSize | number, randomBy: number): number {
+    const base = getSize(size);
+    const random = base * randomBy;
+    const min = base - random;
+    const max = base + random;
+    return Math.floor(Math.random() * (max - min) + min);
+}
+
+function getSize(size: MockGoalSize | number): number {
+    if (typeof size === "number") {
+        return size;
+    } else {
+        return +size;
+    }
+}
+
+function createGoalExecutor(seconds: number, code: number = 0): ExecuteGoal {
+    return async gi => {
+        gi.progressLog.write(`Waiting for ${seconds}s`);
+        await wait(seconds);
+        gi.progressLog.write(`Finished waiting for ${seconds}s`);
+        return {
+            code,
+        };
+    };
+}
+
+function wait(seconds: number): Promise<void> {
+    return new Promise<void>(resolve => {
+        setTimeout(() => resolve(), seconds * 1000).unref();
```

Comment:
What is the purpose of calling `unref()`? Is there a case when there wouldn't be something waiting for this?

Reply:
unref prevents from waiting on this if it wants to shut down.